homeassistant/components/wmspro/config_flow.py

```diff
@@ -0,0 +1,79 @@
+"""Config flow for WMS WebControl pro API integration."""
+
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+import aiohttp
+import voluptuous as vol
+from wmspro.webcontrol import WebControlPro
+
+from homeassistant.components import dhcp
+from homeassistant.config_entries import ConfigFlow, ConfigFlowResult
+from homeassistant.const import CONF_HOST
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import DOMAIN, SUGGESTED_HOST
+
+_LOGGER = logging.getLogger(__name__)
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST, description={"suggested_value": SUGGESTED_HOST}): str,
+    }
+)
+
+
+async def setup_and_verify(hass: HomeAssistant, host: str) -> WebControlPro:
+    """Set up and verify the connection to the wmspro."""
+    session = async_get_clientsession(hass)
+
+    hub = WebControlPro(host, session)
+    try:
+        pong = await hub.ping()
+    except aiohttp.ClientError as err:
+        raise CannotConnect from err
+    if not pong:
+        raise CannotConnect
+
+    return hub
+
+
+class WebControlProConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for wmspro."""
+
+    VERSION = 1
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            try:
+                host = user_input[CONF_HOST]
+                await setup_and_verify(self.hass, host)
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except Exception:
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+            else:
+                return self.async_create_entry(title=host, data=user_input)
+
+        data_schema = STEP_USER_DATA_SCHEMA
+        if self.source == dhcp.DOMAIN:
```

Comment:
Why don't we have a dhcp entry point?

Reply:
Because it is not needed. The DHCP discovered entries need to be user-confirmed anyway according to the developer documentation. And this way the same form can be re-used.

Reply:
Do we get anything unique from the dhcp message?

Reply:
The MAC address indeed, but since the WebControl API does not give us a the MAC address, we cannot match discovered and manually created config_entries.

Reply:
In that case we do want to set the mac address as unique_id and have a separate entrypoint for the DHCP. In that endpoint we should abort if the unique_id is already set. This avoids the device to be shown after being added with discovery.

We should also look into logic that is able to set the unique_id of a config entry that is setup manually whenever we discover that via DHCP

Reply:
> In that case we do want to set the mac address as unique_id and have a separate entrypoint for the DHCP. In that endpoint we should abort if the unique_id is already set. This avoids the device to be shown after being added with discovery.

This is already the case without any extra code as far as I can tell. The hub is not discovered again. But I will test this again and can add the MAC address as unique_id if still needed.

> We should also look into logic that is able to set the unique_id of a config entry that is setup manually whenever we discover that via DHCP

I tried that, but since the hub does not expose any identifying information except the hostname, devices (destinations), rooms and scenes, I wouldn't know a good way to merge/update the discovered and manually setup config_entry this way. As you and the developer documentation state, the hostname shouldn't be used for such purpose.

Reply:
> This is already the case without any extra code as far as I can tell. The hub is not discovered again. But I will test this again and can add the MAC address as unique_id if still needed.

But we still want to set a unique_id whenever possible.

I asked around and integration like `reolink` and `dlink` have the mechanism to attach the unique_id when discovering the device after manual setup

Reply:
Yes, I know, but it compares with existing entries based on their MAC address which it got from the reolink API after manual setup.

Reply:
You will find this part with `host.api.mac_address` all over the place in `reolink`s `config_flow.py`:
https://github.com/home-assistant/core/blob/f9a2408dfa814f5e581b6a121b214a00a933b541/homeassistant/components/reolink/config_flow.py#L243

Reply:
The requirement of the MAC being set after manual config_entry creation can also be derived from this statement in the developer documentation:
> If an integration wants to receive discovery flows to update the IP Address of a device when it comes
online, but a `hostname` or `oui` match would be too broad, **and it has registered in the device registry with mac address using the `CONNECTION_NETWORK_MAC`**,
it should add a DHCP entry with `registered_devices` set to `true`.

https://github.com/home-assistant/developers.home-assistant/blob/3635dae0f59c039764e054d9b0e801b3b409c129/docs/creating_integration_manifest.md?plain=1#L324-L326

Emphasis on "**and it has registered in the device registry with mac address using the `CONNECTION_NETWORK_MAC`**"

Reply:
Right, reolink isn't a good example. `dlink` seems to be one
https://github.com/home-assistant/core/blob/2cc916db6d09b813b158dc5abf96e2b327962dc3/homeassistant/components/dlink/config_flow.py#L27-L42

So if the entry doesn't have an unique_id yet and the host is the same, we assign an unique_id

Reply:
Isn't this unsafe and against recommendations from the developer docs?

IPs could change over time due to DHCP and hostnames are not supposed to be used for this purpose, right?

If you still want to have this added, I would base the matching on the hostname instead of the IP as this is a setting in the hub and not on the network router depending on an optional static IP assignment there.

Reply:
we could check for loaded entries then, we know for sure that the entry works and that the IP works.
I think that works

Reply:
But `CONF_HOST` could be an IP **or** hostname depending on what the user put in initially. So I guess both cases could be handled for loaded entries.

Reply:
oh right, hmmmmm

Reply:
I implemented this now, but dropped the `updates={CONF_HOST: discovery_info.ip}` to not **force** users into using either hostname or IP address. I don't think we should make this decision for our users.